### PR TITLE
Add telemetry app product change message

### PR DESCRIPTION
**Pull request by jandro996**
# What Does This Do

- Add new app_product_change request type
- Update request boy writeProducts method to be able to build the json making each product optional as the schema defines
- Add new Product change telemetry event
- Add more tests

# Motivation

ASM enablement (appsec.enabled) in telemetry product action should reflect remote config changes. Digging into this we noticed that the [app_product_change](https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v2/SchemaDocumentation/Schemas/app_product_change.md) telemetry request type is not implemented in java

# Additional Notes

Only implemented to detect appsec product changes

# Contributor Checklist

- [x] Format the title [according the contribution guidelines](/CONTRIBUTING.md#title-format)
- [x] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [ ] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [APPSEC-53874]

[APPSEC-53874]: https://datadoghq.atlassian.net/browse/APPSEC-53874?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: telemetry/src/main/java/datadog/telemetry/TelemetryRequestBody.java [hunk 1]
```diff
@@ -264,30 +261,58 @@ public void endDependencies() throws IOException {
...
   }
 
+  /**
+   * Use a Pair of booleans to optionally write the products section of the telemetry request. left
+   * element is a boolean indicating if the products section should be written, right element is the
+   * enabling value.
+   */
   public void writeProducts(
-      boolean appsecEnabled, boolean profilerEnabled, boolean dynamicInstrumentationEnabled)
+      Pair<Boolean, Boolean> appsecEnabling,
```
Comment (manuel-alvarez-alvarez) [inline telemetry/src/main/java/datadog/telemetry/TelemetryRequestBody.java:270]: A map product -> status with the products than changed might be more concise and extensible in the future.

File: telemetry/src/main/java/datadog/telemetry/TelemetryService.java [hunk 2]
```diff
@@ -235,6 +248,46 @@ public boolean sendExtendedHeartbeat() {
...
   }
 
+  /** Send app-product-change request with changed product enabled status if any. */
+  public boolean sendAppProductChange() {
+    boolean isAppsecEnabled =
+        Config.get().getAppSecActivation() != ProductActivation.FULLY_DISABLED;
+    boolean isProfilingEnabled = InstrumenterConfig.get().isProfilingEnabled();
+    boolean isDynamicInstrumentationEnabled = Config.get().isDebuggerEnabled();
+
+    boolean hasAppsecEnabledChanged = isAppsecEnabled != this.appsecEnabled;
```
Comment (manuel-alvarez-alvarez) [inline telemetry/src/main/java/datadog/telemetry/TelemetryService.java:258]: Appsec current state should be queried with `AppSecSystem.isActive()`, it's the one that is kept up to date with RC (you might want to check if the other products have something similar).

Comment (jandro996) [reply telemetry/src/main/java/datadog/telemetry/TelemetryService.java:262]: Thanks for the advice! I've changed it to check new value from ActiveSubsystems.APPSEC_ACTIVE that is visible from telemetry

File: telemetry/src/test/groovy/datadog/telemetry/TelemetryServiceSpecification.groovy [hunk 3]
```diff
@@ -486,4 +487,22 @@ class TelemetryServiceSpecification extends DDSpecification {
...
   }
+
+  void 'test app_product_change'() {
+    setup:
+    injectEnvConfig(Strings.toEnvVar(AppSecConfig.APPSEC_ENABLED), "false") // appsec is disabled by default
+
+    TestTelemetryRouter testHttpClient = new TestTelemetryRouter()
+    TelemetryService telemetryService = new TelemetryService(testHttpClient, 10000, false)
+
+    ActiveSubsystems.APPSEC_ACTIVE = true // appsec is enabled by
```
Comment (manuel-alvarez-alvarez) [inline telemetry/src/test/groovy/datadog/telemetry/TelemetryServiceSpecification.groovy:498]:
```suggestion
    ActiveSubsystems.APPSEC_ACTIVE = true // appsec is enabled by RC
```

File: telemetry/src/test/groovy/datadog/telemetry/TelemetryRequestBodySpecification.groovy [hunk 4]
```diff
@@ -107,6 +107,47 @@ class TelemetryRequestBodySpecification extends Specification {
     drainToString(req).contains("\"debug\":true")
   }
 
+  void 'test writeProducts'(){
+    setup:
+    TelemetryRequestBody req = new TelemetryRequestBody(RequestType.APP_PRODUCT_CHANGE)
+    final products = new HashMap<Products, Boolean>()
+    if(appsecChange) {
+      products.put(Products.APPSEC, appsecEnabled)
```
Comment (manuel-alvarez-alvarez) [inline telemetry/src/test/groovy/datadog/telemetry/TelemetryRequestBodySpecification.groovy:115]: You should probably assert that there product is not in the body when no change happened

File: telemetry/src/main/java/datadog/telemetry/TelemetryService.java [hunk 5]
```diff
@@ -235,6 +253,38 @@ public boolean sendExtendedHeartbeat() {
...
+    }
+    log.debug("Preparing app-product-change request");
+    TelemetryRequest request =
+        new TelemetryRequest(
+            this.eventSource,
+            EventSink.NOOP,
+            messageBytesSoftLimit,
+            RequestType.APP_PRODUCT_CHANGE,
+            debug);
+    Map<Products, Boolean> products = new HashMap<>();
```
Comment (manuel-alvarez-alvarez) [inline telemetry/src/main/java/datadog/telemetry/TelemetryService.java:275]: You can use an `EnumMap` instead

**Comment by jandro996** [comment]
Changing the approach to telemetry event queue instead of keep the last product state in memory

File: internal-api/src/main/java/datadog/trace/api/telemetry/Product.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package datadog.trace.api.telemetry;
+
+public class Product {
```
Comment (manuel-alvarez-alvarez) [inline internal-api/src/main/java/datadog/trace/api/telemetry/Product.java:3]: I think `ProductChange` is a better name

Comment (jandro996) [reply internal-api/src/main/java/datadog/trace/api/telemetry/Product.java:3]: This is something that I've doubt a lot when I've changed the approach, I will change it as you suggested

**Review by PerfectSlayer — APPROVED** [review]
Looks good 👍

**Pull request merged**